### PR TITLE
PIL816, PIL-831: Fix/onboarding actions refactor

### PR DESCRIPTION
**Pull request by poocart**
https://linear.app/pillarproject/issue/PIL-816/refactor-onboarding-actions
https://linear.app/pillarproject/issue/PIL-831/improve-offline-onboarding

(part of https://linear.app/pillarproject/issue/PIL-797/maintenance-onboarding-refactor)

Includes:
- Onboading refactor with offline approach and retry fix.
- Other cleanup.
- Marks for suspicious places that is questioned to be removed.

### CODE REVIEW

**Review by graszka22 — COMMENTED** [review]
Is the offline onboarding working? I've tried it and:
- I turn off the internet
- I open the app
- I enter the new username, a bunch of toasts are covering everything:

Screenshot 2020-10-06 at 10 36 43
- I go through the flow, to the home screen, everything is ok
- I turn on the wifi, restart the app
- kaboom, registration failed and crash 

`info: loadSupportedAssetsAction failed: no walletId {"user": {"isEmailVerified": false, "isLegacyUser": true, "isPhoneVerified": false, "username": "angela123"}}`

Screenshot 2020-10-06 at 10 48 14
- even more interesting situation is when the user selects username that is already in the database, the user has no other choice but delete the app because after every restart the app tries to register him with the same username :D  Because of this case I'm wondering if offline onboarding is possible at all 🤔

File: src/actions/walletActions.js [hunk 1]
```diff
@@ -195,44 +84,19 @@ export const generateWalletMnemonicAction = (mnemonicPhrase?: string) => {
...
-    const { appSettings: { data: { themeType } } } = getState();
-    dispatch({
-      type: NEW_WALLET_CONFIRM_PIN,
-      payload: pin,
-    });
-
-    if (shouldRegisterWallet) dispatch(registerWalletAction(false, themeType));
-  };
-};
-
```
Comment (IAmKio) [inline src/actions/walletActions.js:219]: That's a brutal optimisation 😂

File: src/actions/walletActions.js [hunk 2]
```diff
@@ -261,16 +125,9 @@ export const encryptAndSaveWalletAction = (
...
-    const { isImported, isBackedUp, isRecoveryPending } = backupStatus;
+  return async (dispatch: Dispatch) => {
+    dispatch({ type: SET_WALLET_IS_ENCRYPTING, payload: true });
 
-    // might be already in ENCRYPTING state, i.e. on pin change
-    if (walletState !== ENCRYPTING) {
-      dispatch({ type: UPDATE_WALLET_STATE, payload: ENCRYPTING });
-    }
-
-    await delay(50);
```
Comment (IAmKio) [inline src/actions/walletActions.js:273]: Really happy to be seeing these removed 👍

File: src/reducers/userReducer.js [hunk 3]
```diff
@@ -47,15 +46,14 @@ export type UserReducerAction = {
 
 export const initialState: UserReducerState = {
   data: {
+    username: '',
```
Comment (IAmKio) [inline src/reducers/userReducer.js:49]: Better to use something like `null` here?

File: src/screens/ChangePin/ConfirmNewPin/ConfirmNewPin.js [hunk 4]
```diff
@@ -17,120 +17,119 @@
     with this program; if not, write to the Free Software Foundation, Inc.,
     51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */
-import * as React from 'react';
+import React, { useEffect, useState } from 'react';
```
Comment (IAmKio) [inline src/screens/ChangePin/ConfirmNewPin/ConfirmNewPin.js:20]: Ohhhh here we go! 👍

File: src/screens/NewProfile/NewProfile.js [hunk 5]
```diff
@@ -142,32 +139,28 @@ const NewProfile = ({
...
 
-
-  useEffect(() => {
-    if (walletState !== CHECKING_USERNAME && isCheckingUsername) setIsCheckingUsername(false);
-  }, [walletState]);
-
   const colors = getThemeColors(theme);
 
-  const existingUser = !!apiUser?.walletId;
+  const existingUser = !!user?.walletId;
```
Comment (IAmKio) [inline src/screens/NewProfile/NewProfile.js:160]: is `existingUser` worth a `useState`?

File: src/screens/Services/Services.js [hunk 6]
```diff
@@ -252,13 +253,12 @@ class ServicesScreen extends React.Component<Props, State> {
       });
     }
 
-    if (isAltalixEnabled && isAltalixAvailable) {
+    if (isAltalixEnabled && isAltalixAvailable && walletId) {
```
Comment (IAmKio) [inline src/screens/Services/Services.js:256]: Great catches 👍

File: src/translations/locales/en/auth.json [hunk 7]
```diff
@@ -64,7 +64,8 @@
       "default": "Invalid address"
     },
     "tooManyAttemptsTryAgain": "Too many attempts, please try again in {{waitDuration}} seconds.",
-    "missingData": "Please specify the {{missingData}}"
+    "missingData": "Please specify the {{missingData}}",
+    "registrationApiFailedWithNoReason": "Registration API failed with no reason"
```
Comment (IAmKio) [inline src/translations/locales/en/auth.json:68]: Is this something that is displayed to the user? If it is ... we might need to make it a little bit less technical. Maybe:

```suggestion
    "registrationApiFailedWithNoReason": "An unexpected problem has occurred - please contact our support team"
```

**Review by IAmKio — CHANGES_REQUESTED** [review]
🎉  **Absolutely fantastic work given how essential this piece of work is to the maintenance and user experience!** 🎉

Just a few minor points, suggestions, questions and me being pleased with things 😂  Anything you need more information about please let me know!

File: src/actions/__tests__/authActions.test.js [hunk 8]
```diff
@@ -20,22 +20,19 @@
 import configureMockStore from 'redux-mock-store';
 import thunk from 'redux-thunk';
 import ReduxAsyncQueue from 'redux-async-queue';
+import merge from 'lodash.merge';
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/__tests__/authActions.test.js:23]: Good point, it's possible and the only reason I think of why we have it separate is to not enlarge our source code in size, however, I don't think that having it as single will have major impact on size nor having it as single will have any huge benefit either so I don't know honestly if this change is that crucial 😄  Anyway, if we want we can address this as separate ticket as it requires lots of changes.

File: src/actions/__tests__/onboardingActions.test.js [hunk 9]
```diff
@@ -94,163 +91,212 @@ jest.mock('services/api', () => jest.fn().mockImplementation(() => ({
...
+        backupStatus: mockBackupStatus,
+        data: mockImportedWallet,
       },
-      { type: UPDATE_WALLET_STATE, payload: DECRYPTED },
+    });
+
+    const expectedActions = [
+      { type: SET_REGISTERING_USER, payload: true },
+      { type: SET_USER, payload: { username: mockUser.username } },
+      { type: SET_REGISTERING_USER, payload: false },
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/__tests__/onboardingActions.test.js:222]: Yes, because it's setting in state `isRegistering` which is a state when the back-end calls are made. Does this makes sense?

File: src/actions/__tests__/onboardingActions.test.js [hunk 10]
```diff
@@ -94,163 +91,212 @@ jest.mock('services/api', () => jest.fn().mockImplementation(() => ({
...
+        type: UPDATE_ASSETS,
+        payload: { [DEFAULT_ACCOUNTS_ASSETS_DATA_KEY]: transformAssetsToObject(mockInitialAssets) },
+      },
     ];
 
-    return store.dispatch(registerWalletAction())
+    return store.dispatch(setupAppServicesAction())
       .then(() => {
         const actualActions = store.getActions();
         expect(actualActions).toEqual(expectedActions);
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/__tests__/onboardingActions.test.js:302]: You're right, I can check once more if there are any edge cases we should care wrapping into tests 👍

File: src/actions/assetsActions.js [hunk 11]
```diff
@@ -427,25 +426,32 @@ export const fetchAllAccountsBalancesAction = () => {
...
 export const fetchInitialAssetsAction = () => {
   return async (dispatch: Dispatch, getState: GetState, api: SDKWrapper) => {
+    const {
+      user: { data: user },
+      accounts: { data: accounts },
+    } = getState();
+
+    const walletId = user?.walletId;
+    if (!walletId) {
+      reportLog('fetchInitialAssetsAction failed: no walletId', { user });
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/assetsActions.js:436]: 👍

File: src/actions/assetsActions.js [hunk 12]
```diff
@@ -547,9 +553,13 @@ export const searchAssetsAction = (query: string) => {
       return;
     }
 
-    const { user: { data: { walletId } } } = getState();
+    const walletId = user?.walletId;
+    if (!walletId) {
+      reportLog('searchAssetsAction failed: no walletId', { user });
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/assetsActions.js:558]: 👍

File: src/actions/assetsActions.js [hunk 13]
```diff
@@ -596,13 +606,19 @@ export const getAllOwnedAssets = async (api: SDKWrapper, accountId: string, supp
...
+      user: { data: user },
       session: { data: { isOnline } },
     } = getState();
 
     // nothing to do if offline
     if (!isOnline) return;
 
+    const walletId = user?.walletId;
+    if (!walletId) {
+      reportLog('loadSupportedAssetsAction failed: no walletId', { user });
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/assetsActions.js:618]: 👍

File: src/actions/authActions.js [hunk 14]
```diff
@@ -177,20 +177,93 @@ export const loginAction = (
...
-      if (userState === REGISTERED) {
+      const { address } = wallet;
+      let unlockedWallet = { address };
+
+      /**
+       * if user isn't completely registered then put private key into state to complete registration,
+       * this can happen due offline or web recovery portal onboarding
+       */
+      if (!user?.walletId) {
+        unlockedWallet = { ...unlockedWallet, privateKey: decryptedPrivateKey };
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/authActions.js:200]: We had the same before, but in this PR it's just little reorganized. Before we had it for various edge cases, however, if user is registered then this won't happen, it's just replicating part of onboarding state when we keep private key in state as well until registration is complete.

File: src/actions/authActions.js [hunk 15]
```diff
@@ -177,20 +177,93 @@ export const loginAction = (
...
+        dispatch(loadFeatureFlagsAction());
+
+        // to get exchange supported assets in order to show only supported assets on exchange selectors
+        // and show exchange button on supported asset screen only
+        dispatch(getExchangeSupportedAssetsAction());
+
+        // offline onboarded or very old user that doesn't have any smart wallet, let's create one and migrate safe
+        if (!smartWalletAccount) {
+          // this will import and set Smart Wallet as current active account
+          await dispatch(importSmartWalletAccountsAction(decryptedPrivateKey));
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/authActions.js:255]: There are plenty if awaits in our codebase for various dispatch, it's not anti-pattern as even Redux shows examples (i.e. https://redux.js.org/tutorials/essentials/part-5-async-logic#checking-thunk-results-in-components) of ways to do this with `thunk` middleware which we have since the early days and which allows us to have actions as promises.

To answer your question – yes, it's done to wait until action completes.

File: src/actions/onboardingActions.js [hunk 16]
```diff
@@ -18,398 +18,365 @@
...
+    const isRecoveryPending = !!isPortalRecovery;
+    const backupStatus = { isImported, isBackedUp: !!isImported, isRecoveryPending };
+
+    // dispatch to reducer only, will be stored with encryptAndSaveWalletAction
+    dispatch({ type: UPDATE_WALLET_BACKUP_STATUS, payload: backupStatus });
+
+    // encrypt and store
+    await dispatch(encryptAndSaveWalletAction(pinCode, ethersWallet, backupStatus, enableBiometrics));
+
+    dispatch(saveDbAction('app_settings', { appSettings: { wallet: +new Date() } }));
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/onboardingActions.js:190]: It sets wallet creation timestamp.

File: src/actions/badgesActions.js [hunk 17]
```diff
@@ -47,10 +50,16 @@ import { offlineApiCall } from './offlineApiActions';
...
   return async (dispatch: Dispatch, getState: GetState, api: SDKWrapper) => {
     const {
-      user: { data: { walletId } },
+      user: { data: user },
       badges: { data: badges },
     } = getState();
 
+    const walletId = user?.walletId;
+    if (!walletId) {
+      reportLog('fetchBadgesAction failed: no walletId', { user });
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/badgesActions.js:59]: 👍

File: src/actions/badgesActions.js [hunk 18]
```diff
@@ -83,9 +92,16 @@ export const fetchBadgesAction = (notifyOnNewBadge: boolean = true) => {
...
   return async (dispatch: Dispatch, getState: GetState, api: SDKWrapper) => {
     const {
-      user: { data: { walletId } },
+      user: { data: user },
       badges: { contactsBadges },
     } = getState();
+
+    const walletId = user?.walletId;
+    if (!walletId) {
+      reportLog('fetchContactBadgesAction failed: no walletId', { user });
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/badgesActions.js:101]: 👍

File: src/actions/walletActions.js (hunk 1 above)
Comment (poocart) [reply src/actions/walletActions.js:219]: 🔥 🔥 🔥

File: src/actions/walletActions.js (hunk 2 above)
Comment (poocart) [reply src/actions/walletActions.js:273]: 🔥 🔥 🔥

File: src/reducers/userReducer.js (hunk 3 above)
Comment (poocart) [reply src/reducers/userReducer.js:49]: Good point, but this will require a lot of changes in many screens, I will add as separate PR.

File: src/translations/locales/en/auth.json (hunk 7 above)
Comment (poocart) [reply src/translations/locales/en/auth.json:68]: 👍

File: src/screens/NewProfile/NewProfile.js (hunk 5 above)
Comment (poocart) [reply src/screens/NewProfile/NewProfile.js:160]: It will change every time `user` prop changes since `user` prop comes from reducer, component level state management for this not needed.

File: src/actions/onboardingActions.js [hunk 19]
```diff
@@ -18,398 +18,365 @@
...
 
-  const registrationSucceed = !sdkWallet.error && sdkWallet.walletId;
-  const userInfo = registrationSucceed ? await api.userInfo(sdkWallet.walletId) : {};
-  const userState = !isEmpty(userInfo) ? REGISTERED : PENDING;
+    if (isOnline) {
+      api.init();
+      // we us FCM notifications so we must register for FCM, not regular native Push-Notifications
+      await firebaseMessaging.registerForRemoteNotifications().catch(() => null);
+      await firebaseMessaging.requestPermission().catch(() => null);
+      const fcmToken = await firebaseMessaging.getToken().catch(() => null);
```
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/onboardingActions.js:114]: 👍

File: src/actions/__tests__/authActions.test.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/actions/__tests__/authActions.test.js:23]: Definitely not crucial - just through i'd drop this in whilst i saw it 😄 resolving.

File: src/actions/__tests__/onboardingActions.test.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/actions/__tests__/onboardingActions.test.js:222]: Ahh got it, apologies i didn't realise it was a call "trail" as opposed to a literal object. Resolving.

File: src/actions/authActions.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/actions/authActions.js:200]: Okay understood - so actually the user wouldn't even be able to get to the stage where their state is being replicated to Firestore without passing this stage first?

File: src/actions/authActions.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/actions/authActions.js:255]: Super

File: src/actions/onboardingActions.js (hunk 16 above)
[the remark this answers is not in the document]
Comment (IAmKio) [reply src/actions/onboardingActions.js:190]: Cool - it wasn't immediately obvious from the property name `wallet` 👍 i thinking an ID of sorts or an address was supposed to go in there.

File: src/screens/NewProfile/NewProfile.js (hunk 5 above)
Comment (IAmKio) [reply src/screens/NewProfile/NewProfile.js:160]: Cool 👍

File: src/actions/authActions.js (hunk 14 above)
[the remark this answers is not in the document]
Comment (poocart) [reply src/actions/authActions.js:200]: We should filter `wallet` reducer from storing it into Firebase in general.

**Review by IAmKio — APPROVED** [review]
